nemo/collections/tts/models/univnet.py

```diff
@@ -0,0 +1,373 @@
+from nemo.collections.tts.losses.stftlosses import MultiResolutionSTFTLoss
+from nemo.collections.tts.models.base import Vocoder
+from nemo.collections.tts.modules.univnet_modules import MultiPeriodDiscriminator, MultiResolutionDiscriminator
+from nemo.core.classes import Exportable
+from nemo.core.classes.common import PretrainedModelInfo, typecheck
+from nemo.core.neural_types.elements import AudioSignal, MelSpectrogramType
+from nemo.core.neural_types.neural_type import NeuralType
+from nemo.core.optim.lr_scheduler import CosineAnnealing
+from nemo.utils import logging
+
+HAVE_WANDB = True
+try:
+    import wandb
+except ModuleNotFoundError:
+    HAVE_WANDB = False
+
+
+class UnivNetModel(Vocoder, Exportable):
+    def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        if isinstance(cfg, dict):
+            cfg = OmegaConf.create(cfg)
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.audio_to_melspec_precessor = instantiate(cfg.preprocessor)
+        # use a different melspec extractor because:
+        # 1. we need to pass grads
+        # 2. we need remove fmax limitation
+        self.trg_melspec_fn = instantiate(cfg.preprocessor, highfreq=None, use_grads=True)
+        self.generator = instantiate(cfg.generator,
+                                     n_mel_channels=cfg.preprocessor.nfilt, hop_length=cfg.preprocessor.n_window_stride)
+        self.mpd = MultiPeriodDiscriminator(cfg.discriminator.mpd, debug=cfg.debug if "debug" in cfg else False)
+        self.mrd = MultiResolutionDiscriminator(cfg.discriminator.mrd, debug=cfg.debug if "debug" in cfg else False)
+
+        self.discriminator_loss = DiscriminatorLoss()
+        self.generator_loss = GeneratorLoss()
+
+        # reshape MRD resolutions hyperparam and apply them to MRSTFT loss
+        self.stft_resolutions = cfg.discriminator.mrd.resolutions
+        self.fft_sizes = [res[0] for res in self.stft_resolutions]
+        self.hop_sizes = [res[1] for res in self.stft_resolutions]
+        self.win_lengths = [res[2] for res in self.stft_resolutions]
+        self.mrstft_loss = MultiResolutionSTFTLoss(self.fft_sizes, self.hop_sizes, self.win_lengths)
+        self.stft_lamb = cfg.stft_lamb
+
+        self.sample_rate = self._cfg.preprocessor.sample_rate
+        self.stft_bias = None
+
+        if self._train_dl and isinstance(self._train_dl.dataset, MelAudioDataset):
+            self.input_as_mel = True
+        else:
+            self.input_as_mel = False
+
+        self.automatic_optimization = False
+
+    def configure_optimizers(self):
+        self.optim_g = instantiate(self._cfg.optim, params=self.generator.parameters(),)
+        self.optim_d = instantiate(
+            self._cfg.optim, params=itertools.chain(self.mrd.parameters(), self.mpd.parameters()),
+        )
+
+        if hasattr(self._cfg, 'sched'):
+            self.scheduler_g = CosineAnnealing(
```

Comment:
Is it true that Univnet original paper uses CosineAnnealing?

Reply:
it indeed uses the constant lr. Removed `sched` for faithfulness in [235f87b](https://github.com/NVIDIA/NeMo/pull/3186/commits/235f87ba67a76dc0bda24aeff4ed61568fbfc4f3)